File: fern/snippets/tool-router/python/mcp-claude.py
import asyncio
from dotenv import load_dotenv
from claude_agent_sdk.client import ClaudeSDKClient
from claude_agent_sdk.types import (
    ClaudeAgentOptions,
    AssistantMessage,
    TextBlock,
    ToolUseBlock,
)
from composio import Composio

load_dotenv()

# Initialize Composio (API key from env var COMPOSIO_API_KEY or pass explicitly)
composio = Composio()

# Unique identifier of the user
user_id = "user_123"
# Create a tool router session for the user
session = composio.create(user_id=user_id)

# Configure Claude with Composio MCP server
options = ClaudeAgentOptions(
    system_prompt=(
        "You are a helpful assistant with access to external tools. "
        "Always use the available tools to complete user requests."
    ),
    mcp_servers={
        "composio": {
            "type": "http",
            "url": session.mcp.url,
            "headers": session.mcp.headers,  # Authentication headers for the Composio MCP server
        }
    },
    permission_mode="bypassPermissions",  # Auto-approve tools (demo only)
)

async def main():
    print("""
What task would you like me to help you with?
I can use tools like Gmail, GitHub, Linear, Notion, and more.
(Type 'exit' to exit)
Example tasks:
  • 'Summarize my emails from today'
  • 'List all open issues on the composio github repository and create a notion page with the issues'
""")

    async with ClaudeSDKClient(options) as client:
        # Multi-turn conversation with agentic tool calling
        while True:
            user_input = input("\nYou: ").strip()
            if user_input.lower() == "exit":
                break
            
            print("\nClaude: ", end="", flush=True)
            try:
                await client.query(user_input)
                async for msg in client.receive_response():
                    if isinstance(msg, AssistantMessage):
                        for block in msg.content:
                            if isinstance(block, ToolUseBlock):
                                print(f"\n[Using tool: {block.name}]", end="")
                            elif isinstance(block, TextBlock):
                                print(block.text, end="", flush=True)
            except Exception as e:
                print(f"\n[Error]: {e}")

if __name__ == "__main__":
    asyncio.run(main())

File: fern/snippets/tool-router/typescript/mcp-claude.ts
import "dotenv/config"; // Load environment variables from .env file
import { query, type Options } from "@anthropic-ai/claude-agent-sdk";
import { Composio } from "@composio/core";
import { createInterface } from "readline/promises";

const composioApiKey = process.env.COMPOSIO_API_KEY;
if (!composioApiKey) {
  console.warn("⚠️  Warning: COMPOSIO_API_KEY not set - Claude won't be able to use Composio tools");
}
const userId = "user_123"; // Your user's unique identifier

// Initialize Composio and create a Tool Router session
const composio = new Composio({ apiKey: composioApiKey });
const session = await composio.create(userId);

// Configure Claude with Composio MCP server
const options: Options = {
  systemPrompt: "You are a helpful assistant with access to external tools. " +
    "Always use the available tools to complete user requests.",
  mcpServers: {
    composio: { 
      type: "http", 
      url: session.mcp.url, 
      headers: { "x-api-key": composioApiKey } 
    },
  },
  permissionMode: "bypassPermissions",
  allowDangerouslySkipPermissions: true,
};

async function main() {
  console.log("Starting Claude agent with Composio Tool Router...\n");
  
  const rl = createInterface({ input: process.stdin, output: process.stdout });
  
  // Initial task
  const task = "Fetch all open issues from the composio/composio GitHub repository " +
    "and create a Google Sheet with issue number, title, labels, and author";
  
  console.log(`Task: ${task}\n`);
  
  let isFirstQuery = true;
  
  for await (const msg of query({ prompt: task, options })) {
    if (msg.type === "assistant") {
      for (const block of msg.message.content) {
        if (block.type === "tool_use") {
          console.log(`\n[🔧 Using tool: ${block.name}]`);
        } else if (block.type === "text") {
          process.stdout.write(block.text);
        }
      }
    } else if (msg.type === "result" && msg.subtype === "success") {
      console.log(`\n${msg.result}\n`);
    }
  }
  console.log("\n");
  
  isFirstQuery = false;
  
  // If authentication is needed, Claude will provide a link
  console.log("\nOptions:");
  console.log("  - Type 'quit' to exit");
  console.log("  - Say 'connected' or enter a new task\n");
  
  const userInput = await rl.question("You: ");
  
  if (userInput.toLowerCase() !== "quit" && userInput.toLowerCase() !== "exit" && userInput) {
    // Continue the conversation with the same context
    for await (const msg of query({ prompt: userInput, options: { ...options, continue: true } })) {
      if (msg.type === "assistant") {
        for (const block of msg.message.content) {
          if (block.type === "tool_use") {
            console.log(`\n[🔧 Using tool: ${block.name}]`);
          } else if (block.type === "text") {
            process.stdout.write(block.text);
          }
        }
      } else if (msg.type === "result" && msg.subtype === "success") {
        console.log(`\n${msg.result}\n`);
      }
    }
    console.log();
  }
  
  rl.close();
}

main().catch(console.error);